Frameworks/CoreGraphics/CGDataProvider.mm

```diff
@@ -16,130 +16,395 @@
 
 #import <StubReturn.h>
 #import <Starboard.h>
-#import <Foundation/NSData.h>
-#import <Foundation/NSString.h>
-#include "LoggingNative.h"
 #import <CoreGraphics/CGDataProvider.h>
 #import "CGDataProviderInternal.h"
+#import <CFCppBase.h>
 
-static const wchar_t* TAG = L"CGDataProvider";
+// Required so we can use CFHashBytes rather than trying our own hash algorithm
+#import <ForFoundationOnly.h>
 
-// TODO: Task 7188763-Reimplement CGDataProvider which does not inherit from NSData but rather uses it within itself.
-@implementation CGDataProvider : NSData
-- (void)dealloc {
-    if (releaseFunc != 0) {
-        releaseFunc(info, data, size);
+#import <vector>
+
+// Determines what amount of bytes to try and read at a time for indirect access data providers
+static constexpr size_t c_streamSize = 2048;
+
+#pragma region __CGDataProviderInternal
+namespace __CGDataProviderInternal {
+class __CGDataProviderInternal {
+    mutable std::unique_ptr<CFHashCode> m_hashValue;
+
```

Comment:
are you using unique_ptr for std::optional?
Having a 4-byte indirect allocation is a minor waste :)